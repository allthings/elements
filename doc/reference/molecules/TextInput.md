<!-- 
This is an auto-generated markdown. 
You can change it in "src/molecules/TextInput.jsx" and run build:docs to update this file.
-->
# TextInput
TextInputs are used to allow users to enter information like names, numbers, urls, email addresses or passwords.

```example
<TextInput name="email" type="email" placeholder="Your email" required />
<TextInput name="inquiry" lines={5} placeholder="Your question" maxLength={255} minLength={50} />
```
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|backgroundColor|string|Background color of the form item
