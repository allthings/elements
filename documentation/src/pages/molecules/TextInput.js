import React from 'react'
import { TextInput, View } from '@allthings/elements'
import Example from '../../components/Example'
import Notes from '../../components/Notes'

export default () => (
  <View>
    <Notes for={TextInput} />
    <Example>
      {`<TextInput name="email" type="email" placeholder="Your email" required />
<TextInput lines={5} placeholder="Your question" maxLength={255} minLength={50} />`}
    </Example>
  </View>
)
