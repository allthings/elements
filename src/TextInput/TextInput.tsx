import * as React from 'react'
import Theme from '../Theme'
import ListItem from '../ListItem'
import Input, { IInputProps } from '../Input'

/**
 * TextInputs are used to allow users to enter information like names, numbers, urls, email addresses or passwords.
 *
 * ```example
 * <TextInput name="email" type="email" placeholder="Your email" required />
 * <TextInput name="inquiry" lines={5} placeholder="Your question" maxLength={255} minLength={50} />
 * ```
 */
class TextInput extends React.Component<
  IInputProps & { backgroundColor?: string } & React.HTMLAttributes<HTMLElement>
> {
  render() {
    const { backgroundColor = 'background', ...props } = this.props

    return (
      <Theme>
        {({ colorize }) => (
          <ListItem padded={false} backgroundColor={colorize(backgroundColor)}>
            <Input {...props} />
          </ListItem>
        )}
      </Theme>
    )
  }
}

export default TextInput
