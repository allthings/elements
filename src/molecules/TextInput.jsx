import React from 'react'
import View from '../atoms/View'
import ListItem from './List/ListItem'
import { css } from 'glamor'
import Theme from '../behaviour/Theme'
import PropTypes from 'prop-types'
import Relative from '../atoms/Relative'
import Text from '../atoms/Text'

const styles = {
  input: textColor =>
    css({
      height: '50px',
      width: '100%',
      padding: '10px 15px',
      fontSize: '14px',
      border: 0,
      ':invalid:focus': {
        color: 'red',
      },
    }),
  area: (textColor, lines) =>
    css({
      height: `calc(30px*${lines})`,
      width: '100%',
      padding: '10px 15px',
      fontSize: '14px',
      border: 0,
      ':invalid:focus': {
        color: 'red',
      },
    }),
  error: css({
    backgroundColor: '#c1392b',
    position: 'absolute',
    bottom: '100%',
    left: '0',
    zIndex: 40,
    width: '100%',
    padding: '12px 15px 15px',
    borderRadius: '2px',
    boxShadow: '0 1px 2px 1px rgba(0,0,0,.25)',
  }),
  arrow: css({
    position: 'absolute',
    bottom: '-10px',
    left: '50%',
    marginLeft: '-10px',
    display: 'block',
    width: 0,
    height: 0,
    borderLeft: '10px solid transparent',
    borderRight: '10px solid transparent',
    borderTop: '10px solid #c1392b',
  }),
  required: css({
    position: 'absolute',
    right: 10,
  }),
  placeholder: css({
    position: 'absolute',
    bottom: 2,
    right: 15,
  }),
}

const InputError = ({ children, ...props }) => (
  <View {...styles.error} {...props}>
    <Text color="textOnBackground">{children}</Text>
    <View {...styles.arrow} />
  </View>
)

InputError.propTypes = {
  children: PropTypes.string.isRequired,
}

/**
 * TextInputs are used to allow users to enter information like names, numbers, urls, email addresses or passwords.
 *
 * ```example
 * <TextInput name="email" type="email" placeholder="Your email" required />
 * <TextInput lines={5} placeholder="Your question" maxLength={255} minLength={50} />
 * ```
 */
class TextInput extends React.Component {
  static propTypes = {
    /** The default value to put into the component, without making it controlled */
    defaultValue: PropTypes.string,
    /** Indicates that this field is required */
    required: PropTypes.bool,
    /** The name of this input field */
    name: PropTypes.string.isRequired,
    /** Type, can be: 'tel', 'number', 'text', 'url', 'email' */
    type: PropTypes.oneOf(['tel', 'number', 'text', 'url', 'email']),
    /** Called, when the users changes something */
    onChange: PropTypes.func,
    /** The value, makes this component a controlled component */
    value: PropTypes.string,
    /** Can only be used with type=text. Increase to enable multi-line input */
    lines: PropTypes.number,
    /** Regular expression to validate against */
    pattern: PropTypes.string,
    /** Min number of characters that must be provided */
    minLength: PropTypes.number,
    /** Max number of characters that can be provided */
    maxLength: PropTypes.number,
  }

  state = {
    visible: true,
    message: null,
    length: (this.props.value && this.props.value.length) || 0,
  }

  static defaultProps = {
    required: false,
    lines: 1,
    type: 'text',
  }

  handleInvalid = e => {
    e.preventDefault()
    const { STATES, validity } = this.context
    const message = e.target.validationMessage
    let hasState = false
    for (const state of STATES) {
      if (e.target.validity[state]) {
        this.setState({ message: validity[state] || message })
        hasState = true
        break
      }
    }
    hasState && this.setState({ message })
  }

  setInput = input => {
    if (input) {
      this.setState({ length: input.value && input.value.length })
      input.addEventListener('invalid', this.handleInvalid)
    } else {
      this.input.removeEventListener('invalid', this.handleInvalid)
    }
    this.input = input
  }

  handleChange = e => {
    this.setState({ message: null, length: e.target.value.length })
    this.props.onChange && this.props.onChange(e)
  }

  handleMessageClick = () => this.setState({ message: null })

  render() {
    const { required, lines, ...props } = this.props
    return (
      <Theme>
        {({ theme, colorize }) => (
          <ListItem padded={false}>
            <Relative style={{ width: '100%' }}>
              {this.state.message && (
                <InputError onClick={this.handleMessageClick}>
                  {this.state.message}
                </InputError>
              )}
              {required && (
                <View {...styles.required}>
                  <Text color="secondaryText" size="s">
                    *
                  </Text>
                </View>
              )}
              {lines === 1 ? (
                <input
                  ref={this.setInput}
                  {...styles.input(theme.secondaryText)}
                  onInvalid={this.handleInvalid}
                  placeholder="Your text"
                  required={required}
                  aria-required={required}
                  {...props}
                  onChange={this.handleChange}
                />
              ) : (
                <textarea
                  {...styles.area(theme.secondaryText, lines)}
                  {...props}
                  onChange={this.handleChange}
                />
              )}
              {props.maxLength && (
                <View {...styles.placeholder}>
                  <Text color="secondaryText" size="s">
                    {this.state.length}/{props.maxLength}
                  </Text>
                </View>
              )}
            </Relative>
          </ListItem>
        )}
      </Theme>
    )
  }
}

TextInput.contextTypes = {
  validity: PropTypes.object,
  STATES: PropTypes.array,
}

export default TextInput
