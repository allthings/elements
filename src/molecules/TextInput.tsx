import React from 'react'
import Theme from '../behaviour/Theme'
import ListItem from './List/ListItem'
import Input from '../atoms/Input'
import PropTypes from 'prop-types'

/**
 * TextInputs are used to allow users to enter information like names, numbers, urls, email addresses or passwords.
 *
 * ```example
 * <TextInput name="email" type="email" placeholder="Your email" required />
 * <TextInput name="inquiry" lines={5} placeholder="Your question" maxLength={255} minLength={50} />
 * ```
 */
class TextInput extends React.Component {
  static propTypes = {
    ...Input.propTypes,
    /** Background color of the form item */
    backgroundColor: PropTypes.string,
  }

  render() {
    const { backgroundColor, ...props } = this.props

    return (
      <Theme>
        {({ theme, colorize }) => (
          <ListItem padded={false} backgroundColor={colorize(backgroundColor)}>
            <Input {...props} />
          </ListItem>
        )}
      </Theme>
    )
  }
}

export default TextInput
