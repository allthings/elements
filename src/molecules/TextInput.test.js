import React from 'react'
import renderer from 'react-test-renderer'
import TextInput from './TextInput'

test('TextInput renders', () => {
  const tree = renderer
    .create(
      <TextInput
        name="email"
        label="E-Mail"
        type="email"
        placeholder="E-Mail"
        required
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
