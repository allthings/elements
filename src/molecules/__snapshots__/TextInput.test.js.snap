// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders 1`] = `
.css-lwcmpf,
[data-css-lwcmpf] {
  position: absolute;
  left: 15px;
  font-size: 10px;
  opacity: 0;
  transition: all .225s ease-out;
  -webkit-transition: all .225s ease-out;
  -moz-transition: all .225s ease-out;
}

.css-dau5wj,
[data-css-dau5wj] {
  min-height: 50px;
  border-bottom: 1px solid #ecf0f1;
  background-color: white;
}

.css-h7582n,
[data-css-h7582n] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-ouc013,
[data-css-ouc013] {
  display: inline;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 13px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  box-sizing: border-box;
  height: 50px;
  width: 100%;
  padding: 0 15px;
  padding-top: 0;
  transition: all .225s ease-out;
  border: 0;
  -webkit-transition: all .225s ease-out;
  -moz-transition: all .225s ease-out;
}

.css-13azwyo,
[data-css-13azwyo] {
  position: relative;
}

.css-4p3ux5,
[data-css-4p3ux5] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-b05tpj,
[data-css-b05tpj] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 10px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #626262;
}

<div
  data-css-dau5wj=""
  data-css-h7582n=""
>
  <div
    data-css-13azwyo=""
    data-css-4p3ux5=""
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      data-css-4p3ux5=""
      data-css-lwcmpf=""
      style={
        Object {
          "opacity": 0,
          "top": 12,
        }
      }
    >
      <div
        data-css-4p3ux5=""
        data-css-b05tpj=""
      >
        E-Mail
         
        *
      </div>
    </div>
    <input
      aria-required={true}
      data-css-ouc013=""
      name="email"
      onInvalid={[Function]}
      onKeyUp={[Function]}
      placeholder="E-Mail"
      required={true}
      type="email"
    />
  </div>
</div>
`;
